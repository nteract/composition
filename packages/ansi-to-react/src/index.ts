import * as React from "react";
import { ansiToJson, AnserJsonEntry } from "anser";
import { escapeCarriageReturn } from "escape-carriage";

const LINK_REGEX = /(https?:\/\/(?:www\.|(?!www))[^\s\.]+\.[^\s]{2,}|www\.[^\s]+\.[^\s]{2,})/;

/**
 * Converts ANSI strings into JSON output.
 * @name ansiToJSON
 * @function
 * @param {String} input The input string.
 * @return {Array} The parsed input.
 */
function ansiToJSON(input: string) {
  input = escapeCarriageReturn(input);
  return ansiToJson(input, {
    json: true,
    remove_empty: true
  });
}

/**
 * Converts an Anser bundle into a React Node.
 * @param linkify whether links should be converting into clickable anchor tags.
 * @param bundle Anser output.
 */
function convertBundleIntoReact(
  linkify: boolean,
  bundle: AnserJsonEntry,
  key: number
) {
  const style: { backgroundColor?: string; color?: string } = {};
  if (bundle.bg) {
    style.backgroundColor = `rgb(${bundle.bg})`;
  }
  if (bundle.fg) {
    style.color = `rgb(${bundle.fg})`;
  }

  if (!linkify) {
    return React.createElement("span", { style, key }, bundle.content);
  }

  const words = bundle.content.split(" ").reduce(
    (words: React.ReactNode[], word: string, index: number) => {
      // If this isn't the first word, re-add the space removed from split.
      if (index !== 0) {
        words.push(" ");
      }

      // If  this isn't a link, just return the word as-is.
      if (!LINK_REGEX.test(word)) {
        words.push(word);
        return words;
      }

      words.push(
        React.createElement(
          "a",
          {
            key: index,
            href: word,
            target: "_blank"
          },
          `${word}`
        )
      );
      return words;
    },
    [] as React.ReactNode[]
  );
  return React.createElement("span", { style, key }, words);
}

declare type Props = {
  children: string;
  className?: string;
  linkify?: boolean;
};

export default function Ansi(props: Props) {
  return React.createElement(
    "code",
    { className: props.className },
    ansiToJSON(props.children).map(
      convertBundleIntoReact.bind(null, props.linkify)
    )
  );
}
