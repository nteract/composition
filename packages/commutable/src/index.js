/* @flow */

import type { Notebook as v4Notebook } from './v4';

import type {
  ImmutableNotebook,
  JSONType,
} from './types';

type PlaceholderNotebook = {
  nbformat: number,
  nbformat_minor: number,
};

const v4 = require('./v4');

const {
  emptyNotebook,
  emptyCodeCell,
  emptyMarkdownCell,
  appendCell,
  monocellNotebook,
} = require('./structures');

export type Notebook = PlaceholderNotebook & v4Notebook;

function freezeReviver(k: string, v: JSONType): JSONType {
  return Object.freeze(v);
}

// Expected usage of below is
// fromJS(parseNotebook(string|buffer))

function parseNotebook(notebookString: string): Notebook {
  return JSON.parse(notebookString, freezeReviver);
}

function fromJS(notebookJSON: Notebook): ImmutableNotebook {
  if (notebookJSON.nbformat === 4 && notebookJSON.nbformat_minor >= 0) {
    if (Array.isArray(notebookJSON.cells) && typeof notebookJSON.metadata === 'object') {
      return v4.fromJS(notebookJSON);
    }
  }

  if (notebookJSON.nbformat) {
    throw new TypeError(
      `nbformat v${notebookJSON.nbformat}.${notebookJSON.nbformat_minor} not recognized`
    );
  }

  throw new TypeError('This notebook format is not supported');
}

function toJS(immnb: ImmutableNotebook): v4Notebook {
  if (immnb.get('nbformat') === 4 && immnb.get('nbformat_minor') >= 0) {
    return v4.toJS(immnb);
  }
  throw new TypeError('Only notebook format 4 is supported');
}

// Expected usage is stringifyNotebook(toJS(immutableNotebook))
function stringifyNotebook(notebook: v4Notebook): string {
  return JSON.stringify(notebook, null, 2);
}

module.exports = {
  emptyCodeCell,
  emptyMarkdownCell,
  emptyNotebook,
  appendCell,
  monocellNotebook,
  toJS,
  fromJS,
  stringifyNotebook,
};

export type {
  ImmutableNotebook,
};
