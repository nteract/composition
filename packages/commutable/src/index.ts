import { Map as ImmutableMap } from "immutable";
import * as v4 from "./v4";
import * as v3 from "./v3";
import { ImmutableNotebook, JSONType } from "./types";

// .....................................
// API Exports
// Make sure the index.js.flow types stay in sync with this section
//

// from types
export * from "./types";

// from structures
export {
  emptyCodeCell,
  emptyMarkdownCell,
  emptyNotebook,
  monocellNotebook,
  createCodeCell,
  insertCellAt,
  insertCellAfter,
  deleteCell,
  appendCellToNotebook
} from "./structures";

// v4
export {
  StreamOutput,
  Output,
  createImmutableMimeBundle,
  createImmutableOutput
} from "./v4";

// general

export type Notebook = v4.Notebook | v3.Notebook;

const freezeReviver = <T extends JSONType>(k: string, v: T) =>
  Object.freeze(v) as T;

// Expected usage of below is fromJS(parseNotebook(string|buffer))
export const parseNotebook = (notebookString: string): Notebook =>
  JSON.parse(notebookString, freezeReviver);

export const fromJS = (
  notebook: Notebook | ImmutableNotebook
): ImmutableNotebook => {
  if (ImmutableMap.isMap(notebook)) {
    if (notebook.has("cellOrder") && notebook.has("cellMap")) {
      return notebook;
    }
    throw new TypeError(
      `commutable was passed an Immutable.Map structure that is not a notebook`
    );
  }

  if (notebook.nbformat === 4 && notebook.nbformat_minor >= 0) {
    if (
      Array.isArray(notebook.cells) &&
      typeof notebook.metadata === "object"
    ) {
      return v4.fromJS(notebook);
    }
  } else if (notebook.nbformat === 3 && notebook.nbformat_minor >= 0) {
    return v3.fromJS(notebook);
  }

  if (notebook.nbformat) {
    throw new TypeError(
      `nbformat v${notebook.nbformat}.${notebook.nbformat_minor} not recognized`
    );
  }

  throw new TypeError("This notebook format is not supported");
};

export const toJS = (immnb: ImmutableNotebook): v4.Notebook => {
  const minorVersion: null | number = immnb.get("nbformat_minor", null);

  if (
    immnb.get("nbformat") === 4 &&
    typeof minorVersion === "number" &&
    minorVersion >= 0
  ) {
    return v4.toJS(immnb);
  }
  throw new TypeError("Only notebook formats 3 and 4 are supported!");
};

// Expected usage is stringifyNotebook(toJS(immutableNotebook))
export const stringifyNotebook = (notebook: v4.Notebook) =>
  JSON.stringify(notebook, null, 2);
