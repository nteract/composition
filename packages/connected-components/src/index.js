// @flow

export {
  default as NewNotebookNavigation,
  PureNewNotebookNavigation
} from "./new-notebook-navigation";

export { default as NotebookMenu } from "./notebook-menu";
export { default as ModalController } from "./modal-controller";
