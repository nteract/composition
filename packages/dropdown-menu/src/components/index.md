The dropdown menu used for cell actions is composed of three components that work with each other:

- `<DropdownMenu />` The outermost component
- `<DropdownTrigger />` The Clickable area that opens the menu
- `<DropdownContent />` The UI elements. Each clickable entry should be in an `<li>`

NOTE: This interface is pretty rough, so you'll have to excuse some strangeness for now. ðŸ˜¬

```jsx
const { DropdownMenu, DropdownTrigger, DropdownContent } = require("../");

const triggerStyle = {
  backgroundColor: "#F5D541",
  color: "#AC8EBC",
  padding: "10px",
  margin: "0px",
  width: "100px"
};

const itemStyle = {
  backgroundColor: "#AC8EBC",
  color: "#F5D541"
};

<div>
  <DropdownMenu>
    <DropdownTrigger>
      <p style={triggerStyle}>Drop It</p>
    </DropdownTrigger>
    <DropdownContent>
      <li style={itemStyle} onClick={() => console.log("Try it with me")}>
        Pop It
      </li>
      <li style={itemStyle} onClick={() => console.log("Here we go")}>
        Lock It
      </li>
      <li style={itemStyle} onClick={() => console.log("Boom boom clap")}>
        Polka Dot It
      </li>
    </DropdownContent>
  </DropdownMenu>
</div>;
```

The primary purpose of this component is to have some hidden option to take as actions. Within the cell UI, this forms the cell actions area.
