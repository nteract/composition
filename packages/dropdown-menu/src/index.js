// @flow
/* eslint jsx-a11y/no-static-element-interactions: 0 */
/* eslint jsx-a11y/click-events-have-key-events: 0 */

// react-hot-loader uses proxies to the original elements so we need to use
// their comparison function in case a consumer of these components is
// using hot module reloading
import { areComponentsEqual } from "react-hot-loader";
import * as React from "react";

type DropdownMenuProps = {
  children: React.ChildrenArray<React.Element<*>>
};

type DropdownMenuState = {
  menuHidden: boolean
};

export class DropdownMenu extends React.Component<
  DropdownMenuProps,
  DropdownMenuState
> {
  constructor(props: *) {
    super(props);
    this.state = {
      menuHidden: true
    };
  }

  render() {
    return (
      <div className="dropdown">
        {React.Children.map(this.props.children, child => {
          if (areComponentsEqual(child.type, DropdownTrigger)) {
            return React.cloneElement(child, {
              onClick: ev => {
                this.setState({ menuHidden: !this.state.menuHidden });
              }
            });
          } else if (areComponentsEqual(child.type, DropdownContent)) {
            if (this.state.menuHidden) {
              return null;
            } else {
              // DropdownContent child will pass down an onItemClick so that
              // the menu will collapse
              return React.cloneElement(child, {
                onItemClick: ev => {
                  this.setState({ menuHidden: true });
                }
              });
            }
          } else {
            // fallback
            return child;
          }
        })}
        <style jsx>{`
          .dropdown {
            display: inline-block;
          }
        `}</style>
      </div>
    );
  }
}

export class DropdownTrigger extends React.Component<{
  children: React.ChildrenArray<React.Element<*>>,
  onClick?: (ev: SyntheticEvent<*>) => void
}> {
  render() {
    return (
      <div onClick={this.props.onClick}>
        {this.props.children}
        <style jsx>{`
          div {
            user-select: none;
            margin: 0px;
            padding: 0px;
          }
        `}</style>
      </div>
    );
  }
}

export class DropdownContent extends React.Component<{
  children: React.ChildrenArray<React.Element<*>>,
  onItemClick: (ev: SyntheticEvent<*>) => void
}> {
  static defaultProps = {
    // Completely silly standalone, because DropdownMenu injects the onItemClick handler
    onItemClick: () => {}
  };

  render() {
    return (
      <div>
        <ul>
          {React.Children.map(this.props.children, child => {
            return React.cloneElement(child, {
              onClick: ev => {
                child.props.onClick(ev);
                // Hide the menu
                this.props.onItemClick(ev);
              }
            });
          })}
        </ul>
        <style jsx>{`
          div {
            user-select: none;
            margin: 0px;
            padding: 0px;

            width: 200px;

            opacity: 1;
            position: absolute;
            top: 0.2em;
            right: 0;
            border-style: none;
            padding: 0;
            font-family: var(--nt-font-family-normal);
            font-size: var(--nt-font-size-m);
            line-height: 1.5;
            margin: 20px 0;
            background-color: var(--theme-cell-menu-bg);
          }

          ul {
            list-style: none;
            text-align: left;
            padding: 0;
            margin: 0;
            opacity: 1;
          }

          :global(li) {
            padding: 0.5rem;
          }

          :global(li:hover) {
            background-color: var(--theme-cell-menu-bg-hover, #e2dfe3);
            cursor: pointer;
          }
        `}</style>
      </div>
    );
  }
}
