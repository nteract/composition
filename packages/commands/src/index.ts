import { COMMANDS } from "./commands";
import { makeCommandSetsComponent } from "./components";
import { CellTarget, CommandLocation, GapTarget, makeCellInfo, makeCellModelFromState, makeGapInfo, makeGapModel } from "./model";

export { COMMANDS };
export { CellTarget, CommandLocation, GapTarget, makeCellInfo, makeCellModelFromState, makeGapInfo, makeGapModel };

export const CellCommands =
  makeCommandSetsComponent(makeCellInfo, makeCellModelFromState, COMMANDS.CELL);

export const GapCommands =
  makeCommandSetsComponent(makeGapInfo, makeGapModel, COMMANDS.GAP);
