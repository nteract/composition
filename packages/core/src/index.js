//@flow

import * as actions from "./actions";
import * as constants from "./constants";
import * as middlewares from "./middlewares";
import * as records from "./records";
import * as reducers from "./reducers";
import * as components from "./components";

export { actions, constants, middlewares, records, reducers, components };
