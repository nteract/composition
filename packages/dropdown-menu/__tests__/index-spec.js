// @flow
/* eslint jsx-a11y/no-static-element-interactions: 0 */
/* eslint jsx-a11y/click-events-have-key-events: 0 */
/* eslint jsx-a11y/no-noninteractive-element-interactions: 0 */

import * as React from "react";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";

import { DropdownMenu, DropdownContent, DropdownTrigger } from "../src";

describe("DropdownMenu", () => {
  test("clicking dropdown content triggers the items callback and closes the menu", () => {
    const clicky = jest.fn();

    const wrapper = mount(
      <DropdownMenu>
        <DropdownTrigger>
          <div className="clickMe">Click me</div>
        </DropdownTrigger>
        <DropdownContent>
          <li className="alsoClickMe" onClick={clicky}>
            1
          </li>
        </DropdownContent>
      </DropdownMenu>
    );

    expect(toJSON(wrapper)).toMatchSnapshot();

    // Trigger should be shown
    // Content should not be available yet
    expect(toJSON(wrapper)).toMatchSnapshot();
    expect(wrapper.find(".alsoClickMe").length).toEqual(0);

    // Content should now be shown
    wrapper.find(".clickMe").simulate("click");
    expect(toJSON(wrapper)).toMatchSnapshot();
    expect(wrapper.find(".alsoClickMe").length).toEqual(1);

    // Clicking the item should close the menu
    wrapper.find(".alsoClickMe").simulate("click");
    expect(wrapper.find(".alsoClickMe").length).toEqual(0);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
