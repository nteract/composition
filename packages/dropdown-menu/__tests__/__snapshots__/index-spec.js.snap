// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenu clicking dropdown content triggers the items callback and closes the menu 1`] = `
<DropdownMenu>
  <div
    className="jsx-3444321073 dropdown"
  >
    <DropdownTrigger
      key=".0"
      onClick={[Function]}
    >
      <div
        className="jsx-3273030785"
        onClick={[Function]}
      >
        <div
          className="clickMe"
        >
          Click me
        </div>
        <JSXStyle
          css="div.jsx-3273030785{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;margin:0px;padding:0px;}"
          styleId="3273030785"
        />
      </div>
    </DropdownTrigger>
    <JSXStyle
      css=".dropdown.jsx-3444321073{z-index:10000;display:inline-block;}"
      styleId="3444321073"
    />
  </div>
</DropdownMenu>
`;

exports[`DropdownMenu clicking dropdown content triggers the items callback and closes the menu 2`] = `
<DropdownMenu>
  <div
    className="jsx-3444321073 dropdown"
  >
    <DropdownTrigger
      key=".0"
      onClick={[Function]}
    >
      <div
        className="jsx-3273030785"
        onClick={[Function]}
      >
        <div
          className="clickMe"
        >
          Click me
        </div>
        <JSXStyle
          css="div.jsx-3273030785{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;margin:0px;padding:0px;}"
          styleId="3273030785"
        />
      </div>
    </DropdownTrigger>
    <JSXStyle
      css=".dropdown.jsx-3444321073{z-index:10000;display:inline-block;}"
      styleId="3444321073"
    />
  </div>
</DropdownMenu>
`;

exports[`DropdownMenu clicking dropdown content triggers the items callback and closes the menu 3`] = `
<DropdownMenu>
  <div
    className="jsx-3444321073 dropdown"
  >
    <DropdownTrigger
      key=".0"
      onClick={[Function]}
    >
      <div
        className="jsx-3273030785"
        onClick={[Function]}
      >
        <div
          className="clickMe"
        >
          Click me
        </div>
        <JSXStyle
          css="div.jsx-3273030785{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;margin:0px;padding:0px;}"
          styleId="3273030785"
        />
      </div>
    </DropdownTrigger>
    <DropdownContent
      key=".1"
      onItemClick={[Function]}
    >
      <div
        className="jsx-104434278"
      >
        <ul
          className="jsx-104434278"
        >
          <li
            className="alsoClickMe"
            key=".0"
            onClick={[Function]}
          >
            1
          </li>
        </ul>
        <JSXStyle
          css="div.jsx-104434278{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;margin:0px;padding:0px;width:200px;opacity:1;position:absolute;top:0.2em;right:0;border-style:none;padding:0;font-family:var(--nt-font-family-normal);font-size:var(--nt-font-size-m);line-height:1.5;margin:20px 0;background-color:var(--theme-cell-menu-bg);}ul.jsx-104434278{list-style:none;text-align:left;padding:0;margin:0;opacity:1;}li{padding:0.5rem;}li:hover{background-color:var(--theme-cell-menu-bg-hover,#e2dfe3);cursor:pointer;}"
          styleId="104434278"
        />
      </div>
    </DropdownContent>
    <JSXStyle
      css=".dropdown.jsx-3444321073{z-index:10000;display:inline-block;}"
      styleId="3444321073"
    />
  </div>
</DropdownMenu>
`;

exports[`DropdownMenu clicking dropdown content triggers the items callback and closes the menu 4`] = `
<DropdownMenu>
  <div
    className="jsx-3444321073 dropdown"
  >
    <DropdownTrigger
      key=".0"
      onClick={[Function]}
    >
      <div
        className="jsx-3273030785"
        onClick={[Function]}
      >
        <div
          className="clickMe"
        >
          Click me
        </div>
        <JSXStyle
          css="div.jsx-3273030785{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;margin:0px;padding:0px;}"
          styleId="3273030785"
        />
      </div>
    </DropdownTrigger>
    <JSXStyle
      css=".dropdown.jsx-3444321073{z-index:10000;display:inline-block;}"
      styleId="3444321073"
    />
  </div>
</DropdownMenu>
`;
